### PR TITLE
maint: Docker tagging updates

### PR DESCRIPTION
**Pull request by RainofTerra**
## Which problem is this PR solving?
#790 
-

## Short description of the changes
- For internal dev builds, the version will be the build # from circle. They will also be tagged with this build number
- The tags for dev builds will include `branch-branchname` such as `branch-terra.myawesomebranch`
- For internal dev builds for `main` will be tagged with latest
- For releases, we will tag with `major`, `major.minor`, and `semver` in addition to `latest` (for compatibility) such as `2`, `2.1`, and `2.1.5`
- All dev builds will now be available in our ECR repo, I will separately add a lifecycle policy to it so it only keeps so many days of non-release builds (90 maybe?)

### CODE REVIEW

File: build-docker.sh [hunk 1]
```diff
@@ -1,14 +1,42 @@
...
-    # append version to image tags
-    TAGS+=",$VERSION"
+
+    # Extract major, major.minor, and major.minor.patch versions
+    MAJOR_VERSION=${VERSION%%.*}
+    MINOR_VERSION=${VERSION%.*}
+
+    # Append versions to image tags
+    # So 2.1.1 would be tagged with "2","2.1","2.1.1"
+    TAGS="$MAJOR_VERSION,$MINOR_VERSION,$VERSION,latest"
```
Comment (TylerHelmuth) [inline build-docker.sh:39]: Do all the places we publish images handle new images with previously used tags?

File: build-docker.sh [hunk 2]
```diff
@@ -1,14 +1,42 @@
...
+# We only want to tag main with latest on ECR
+if [[ "${CIRCLE_BRANCH}" == "main" ]]; then
+    TAGS+=",latest"
+fi
+
+# Local builds just get "dev" for version
 VERSION="dev"
+
+# If we are doing a dev build on circle, we will version it as the circleci buildnumber
+if [[ -n "${CIRCLE_BUILD_NUM}" ]]; then
```
Comment (TylerHelmuth) [inline build-docker.sh:21]: Will CIRCLE_BUILD_NUM ever not be set?  Does this case render `VERSION="dev"` obsolete?

Comment (RainofTerra) [reply build-docker.sh:21]: I wasn't sure how people do local builds - do those go through circle?

File: build-docker.sh (hunk 1 above)
Comment (RainofTerra) [reply build-docker.sh:39]: Yeah, the latest by time stamp is always chosen when there are multiples of a tag for both dockerhub and ECR (and everything else I'm aware of)

File: build-docker.sh (hunk 2 above)
Comment (TylerHelmuth) [reply build-docker.sh:21]: You're right, you could technically run this script locally.

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**